drivers/i2c/i2c_andes_atciic100.c

```diff
@@ -0,0 +1,983 @@
+	case 0x0:
+		i2c_info->fifo_depth = 2;
+		break;
+	case 0x1:
+		i2c_info->fifo_depth = 4;
+		break;
+	case 0x2:
+		i2c_info->fifo_depth = 8;
+		break;
+	case 0x3:
+		i2c_info->fifo_depth = 16;
+		break;
+	}
+
+	/*
+	 * I2C setting: slave mode(default), standard speed
+	 * 7-bit, CPU mode
+	 */
+	OUTWORD(I2C_SET(dev), 0x0);
+	reg = INWORD(I2C_SET(dev));
+	reg |= ((SETUP_T_SUDAT_STD << 24)	|
+		(SETUP_T_SP_STD  << 21)		|
+		(SETUP_T_HDDAT_STD << 16)	|
+		(SETUP_T_SCL_RATIO_STD << 13)	|
+		(SETUP_T_SCLHI_STD << 4)	|
+		SETUP_MASTER			|
+		SETUP_I2C_EN);
+
+	OUTWORD(I2C_SET(dev), reg);
+
+	/* Enable completion & address hit interrupt */
+	reg = INWORD(I2C_INTE(dev));
+	reg |= (IEN_CMPL | IEN_ADDR_HIT);
+	OUTWORD(I2C_INTE(dev), reg);
+
+	i2c_info->driver_state = I2C_DRV_INIT;
+	i2c_info->status.mode = 1;
+	i2c_info->status.arbitration_lost = 0;
+	i2c_info->status.bus_error = 0;
+}
+
+static int i2c_atciic100_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_atciic100_device_config *dev_cfg = DEV_I2C_CFG(dev);
+	struct _i2c_info *i2c_info = dev_cfg->i2c_info;
+	uint32_t reg = 0;
+
+	if (k_sem_take(&i2c_info->i2c_transfer_sem, K_MSEC(1000))) {
+		return DRIVER_ERROR_TIMEOUT;
+	}
+
+	if (k_sem_take(&i2c_info->i2c_busy_sem, K_MSEC(1000))) {
+		return DRIVER_ERROR_TIMEOUT;
+	}
+
+	reg = INWORD(I2C_SET(dev));
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		reg |= SETUP_SPEED_STD;
+		break;
+
+	case I2C_SPEED_FAST:
+		reg |= SETUP_SPEED_FAST;
+		break;
+
+	case I2C_SPEED_FAST_PLUS:
+		reg |= SETUP_SPEED_FAST_PLUS;
+
+	case I2C_SPEED_HIGH:
+		return DRIVER_ERROR_UNSUPPORTED;
+	case 0x00:
+		break;
+	default:
+		return DRIVER_ERROR_PARAMETER;
+	}
+
+	if ((dev_config & I2C_MODE_MASTER) == I2C_MODE_MASTER) {
+		reg |= SETUP_MASTER;
+		i2c_info->status.mode = 1;
+	} else {
+		reg &= ~SETUP_MASTER;
+		i2c_info->status.mode = 0;
+	}
+
+	if ((dev_config & I2C_ADDR_10_BITS) == I2C_ADDR_10_BITS) {
+		reg |= SETUP_ADDRESSING;
+	} else {
+		reg &= ~SETUP_ADDRESSING;
+	}
+
+	OUTWORD(I2C_SET(dev), reg);
+
+	i2c_info->driver_state |= I2C_DRV_CFG_PARAM;
+
+	k_sem_give(&i2c_info->i2c_transfer_sem);
+	k_sem_give(&i2c_info->i2c_busy_sem);
+
+	return DRIVER_OK;
+}
+
+static int i2c_atciic100_get_config(const struct device *dev, uint32_t *dev_config)
+{
+	const struct i2c_atciic100_device_config *dev_cfg = DEV_I2C_CFG(dev);
+	uint32_t reg, config = 0;
+
+	if (dev_cfg->i2c_info == NULL) {
+		return -EIO;
+	}
+
+	reg = INWORD(I2C_SET(dev));
+
+	if (reg & SETUP_MASTER) {
+		config |= I2C_MODE_MASTER;
+	}
+
+	switch (reg & SETUP_SPEED_MSK) {
+	case SETUP_SPEED_STD:
+		config |= I2C_SPEED_SET(I2C_SPEED_STANDARD);
+		break;
+
+	case SETUP_SPEED_FAST:
+		config |= I2C_SPEED_SET(I2C_SPEED_FAST);
+		break;
+
+	case SETUP_SPEED_FAST_PLUS:
+		config |= I2C_SPEED_SET(I2C_SPEED_FAST_PLUS);
+
+	default:
+		return -EIO;
+	}
+
+	*dev_config = config;
+
+	return 0;
+}
+
+static int i2c_atciic100_transfer(const struct device *dev,
```

Comment:
I would prefer to see a single exit from this function with k_sem_give there avoiding the potential for missing a k_sem_give as the code evolves.